### ISSUE TITLE
Impossible to use clang to build rust packages

### ISSUE THREAD

**Issue by ninegua** [NixOS/nixpkgs#303796]
Can we just say that use `setEnv` to force set environment variables before the cargo command is a bad idea?

https://github.com/NixOS/nixpkgs/blob/ac6c93e57592408c7fb65cfd35b3293435589089/pkgs/build-support/rust/lib/default.nix#L57

https://github.com/NixOS/nixpkgs/blob/ac6c93e57592408c7fb65cfd35b3293435589089/pkgs/build-support/rust/hooks/cargo-build-hook.sh#L33

It forces `HOST_CC` to be set to `gcc` even when `stdenv` is switched to llvm, for example the code below would still run `gcc` when building native packages because of the `HOST_CC` setting, which would lead to linking failures:

```
makeRustPlatform {
        rustc = self.rust-stable;
        cargo = self.rust-stable;
        stdenv = self.llvmPackages.libcxxStdenv;
}
```

Note that this problem does not exist with Nixpkgs 23.05. It only becomes a problem after the `setEnv` approach is adopted in 23.11 and newer.

In general, we want environment variables to be configurable. But this `setEnv` thing is just very difficult to override. Would appreciate suggestions on how to override `HOST_CC`, or getting rid of `setEnv` completely. Thanks!

**Comment by RossComputerGuy** [comment]
Is this still a problem? Have you tried `pkgsLLVM`?

**Comment by ninegua** [comment]
It is still a problem. It is pretty clear where the problem is, namely, `setEnv` is too rigid and can't be overridden.

**Comment by RossComputerGuy** [comment]
Ok, how have you tried setting the CC? Have you tried overriding rustc's stdenv to `clangStdenv`?

**Comment by ninegua** [comment]
> Ok, how have you tried setting the CC? Have you tried overriding rustc's stdenv to clangStdenv?

If you read the source code of `setEnv`, it gets CC from environment, but it gets `HOST_CC` from `pkgsBuildHost.stdenv.cc`, which is not something a user can easily have control over.

**Comment by RossComputerGuy** [comment]
I am reading the source code which is why I have been asking if you have tried changing the stdenv.

**Comment by ninegua** [comment]
Yes, I tried to pass a different stdenv as argument to `makeRustPlatform`, but it wasn't effective.

**Comment by RossComputerGuy** [comment]
`makeRustPlatform` isn't what needs to have stdenv overridden. `pkgs/development/compilers/rust/default.nix` imports `pkgs/build-support/rust/lib/default.nix` which the rust lib takes in stdenv. Which the rust default file is imported from `pkgs/development/compilers/rust/1_80.nix`.

**Comment by ninegua** [comment]
I don't know, but using `pkgsLLVM` is too much of a heavy lifting to solve a local environment variable problem. But thanks for the suggestion though!

**Comment by RossComputerGuy** [comment]
Yeah, these are two of the best (probably only) options. I suggest `pkgsLLVM` since it gives you an environment that is purely LLVM on target. Overriding a single package to use LLVM will result in an LLVM + GCC environment. If you have any issues with `clangStdenv` when doing this, I suggest giving `pkgsLLVM` a shot.

**Comment by emilazy** [comment]
The Rust build infrastructure should be fixed, though.

**Comment by RossComputerGuy** [comment]
@emilazy True, `pkgs/development/compilers/rust/default.nix` could be `callPackage`d and `pkgs/development/compilers/rust/1_80.nix` could just inherit it and pass through `stdenv`.

**Comment by reckenrode** [comment]
Rust and `setEnv` is what motivated https://github.com/NixOS/nixpkgs/pull/346043. Prior to that PR, it was impossible to override the SDK version of Rust packages that needed to build non-Rust code (particularly some -sys crates) because `setEnv` pulled in other clangs with the wrong SDK version (e.g., https://github.com/NixOS/nixpkgs/issues/239384).

Some of my early attempts to work around it involved turning `makeRustPlatform` into a fixed-point and allowing `rustPlatform` to be overridden, but it uses `pkgsBuildHost`, `pkgsBuildTarget`, and `pkgsTargetTarget` depending on the cross-compilation scenario. Just overriding the `stdenv` may not be enough.
